src/systems/shadow-system.js

```diff
@@ -90,46 +88,44 @@ function resizeShadowCameraFrustum(light, boundingBox) {
 
 export class ShadowSystem {
   constructor(sceneEl) {
-    this.needsUpdate = false;
-    this.dynamicShadowsEnabled = window.APP.store.state.preferences.enableDynamicShadows;
     this.sceneEl = sceneEl;
-    this.onEnvironmentSceneLoaded = this.onEnvironmentSceneLoaded.bind(this);
-    this.sceneEl.addEventListener("environment-scene-loaded", this.onEnvironmentSceneLoaded);
     this.shadowCameraBoundingBox = new THREE.Box3();
+    this.previousShadowsEnabled = this.realtimeShadowsEnabled = null;
+    window.APP.store.addEventListener("statechanged", this.updatePrefs.bind(this));
+    this.updatePrefs();
   }
 
-  onEnvironmentSceneLoaded({ detail: environmentObject3D }) {
-    this.environmentObject3D = environmentObject3D;
-    this.needsUpdate = true;
-    this.sceneEl.renderer.shadowMap.autoUpdate = this.dynamicShadowsEnabled;
+  updatePrefs() {
+    this.realtimeShadowsEnabled =
+      window.APP.store.state.preferences.enableDynamicShadows === true &&
+      window.APP.store.materialQualitySetting !== "low";
+    if (this.previousShadowsEnabled !== this.realtimeShadowsEnabled) {
+      this.sceneEl.setAttribute("shadow", { enabled: this.realtimeShadowsEnabled });
+
+      // If scene has already rendered, materials which can receive shadows must be updated.
+      if (this.sceneEl.hasLoaded) {
+        this.sceneEl.object3D.traverse(node => {
+          if (node.receiveShadow && node.material) {
+            const materials = Array.isArray(node.material) ? node.material : [node.material];
```

Comment:
Fairly minor but in three-utils.js we have `updateMaterials` you could use here.

Reply:
Updated, thanks